### ISSUE TITLE
CCNode tests reveal some oddities

### ISSUE THREAD

**Issue by andykorth** [cocos2d/cocos2d-objc#420]
TLDR:  Andy was very confused, now he's less confused.

---

I've been writing unit tests for CCNodes and found some odd behavior around CCNode position types, contentSizes, UIScale, and the node's transform.

I'm not sure what of this is intended behavior, so everyone's thoughts opinions are appreciated!

Edit: Scott showed me SpriteBuilder, and I realized that many of these fields make sense when you're looking at the values in the inspector as you drag things around the screen. Without having seen SpriteBuilder, I was struggling to figure out what each enumeration did and why.

I've cataloged the failing tests and some inline questions here in my branch:
https://github.com/andykorth/cocos2d-iphone/tree/develop-v3

https://github.com/andykorth/cocos2d-iphone/blob/f8dd1aef0fbdd01a48af2b8f0dbc18cd03a8c87f/UnitTests/CCNodeTests.m#L206
When changing a node's position to be relative to a different corner. (TopRight instead of BottomLeft, the `node.positionInPoints` changes, but the `node.position` does not. At first, I didn't expect the positionInPoints to include changes to the reference corner. Shortly after that, I was confused as to why the positionInPoints and position were measured relative to _different_ corners. (It's measuring in two different units, but also measuring from different places. The different units were expected, given the names, but not that they were measuring from different places). 

https://github.com/andykorth/cocos2d-iphone/blob/f8dd1aef0fbdd01a48af2b8f0dbc18cd03a8c87f/UnitTests/CCNodeTests.m#L247
Changing the `UIScaleFactor` changes the `positionInPoints`, as expected, but the `contentSizeInPoints` is unchanged. I didn't expect that. Scott explained what was going on, I guess now it makes sense. 

https://github.com/andykorth/cocos2d-iphone/blob/f8dd1aef0fbdd01a48af2b8f0dbc18cd03a8c87f/UnitTests/CCNodeTests.m#L321
Changing the node's scale is reflected correctly in the transform, but the `node.contentSize` is unchanged. This may or may not be intentional, but it didn't look like code that was using the contentSize was taking scale into account. 

Scott said the contentSize was in node-local coordinates and shouldn't change when the scale changed. 

Bigger questions:
UIScaleFactor seems to be sort of odd for a few reasons. It's a global scale factor that seems to be applied to everything, (not necessarily just UI stuff). It's clear to me that the UIScaleFactor is the conversion between "normal coordinates" (node.position) and "points" (node.positionInPoints). In other places, it's the difference between "Points" and "UIPoints" (such as convertContentSizeToPoints).

Edit: Okay, after more investigation, it seems to be applied only to positions when asking for (or on a node where node.positionType is set to) CCPositionUnitScaled. It's only applied to scales (and content sizes) when CCSizeUnitScaled is involved. 

https://github.com/andykorth/cocos2d-iphone/blob/f8dd1aef0fbdd01a48af2b8f0dbc18cd03a8c87f/UnitTests/CCNodeTests.m#L288

I'm slowly figuring out what's going on here, and it still seems odd to change the internal representation of the position based on the positionType. There is a LOT of code associated with these scale factors and position types. It mostly seems well encapsulated, but it also seems like some values aren't handled properly, like contentSize.

Edit 2: After seeing cocosBuilder in action, I understand things a bit better. Although I would have expected the node.position value not to change when the positionType changes. Instead, I think the positionType should be used to extract position relative to certain corners or certain scales. If you want to set or get the position relative to the TopRight, you should be using the get or set methods and passing the positionType as an argument. 

Anyway, I'm confused and I don't have a dozen nodes set up with different position types. I assume others will be confused too. 

---

Proposals: I'll flesh out the docs on UIScaleFactor, CCNode's positionType, and scaleType. Make sure people realize that positionType and scaleType operate independently and what each of the two operate on.

**Comment by andykorth** [comment]
I think I've worked out all the issues.
